Source/WebCore/loader/archive/cf/LegacyWebArchiveCallbackAggregator.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <WebCore/FrameIdentifier.h>
+#include <wtf/CompletionHandler.h>
+
+namespace WebCore {
+
+class LegacyWebArchiveCallbackAggregator final : public ThreadSafeRefCounted<LegacyWebArchiveCallbackAggregator, WTF::DestructionThread::MainRunLoop> {
+public:
+    using Callback = CompletionHandler<void(RefPtr<LegacyWebArchive>&&)>;
+    static Ref<LegacyWebArchiveCallbackAggregator> create(FrameIdentifier rootFrameIdentifier, HashMap<FrameIdentifier, Ref<WebCore::LegacyWebArchive>>&& frameArchives, Callback&& callback)
+    {
+        return adoptRef(*new LegacyWebArchiveCallbackAggregator(rootFrameIdentifier, WTFMove(frameArchives), WTFMove(callback)));
+    }
+
+    RefPtr<LegacyWebArchive> completeFrameArchive(FrameIdentifier identifier)
+    {
+        RefPtr archive = m_frameArchives.take(identifier);
+        if (!archive)
+            return archive;
+
+        for (auto subframeIdentifier : archive->subframeIdentifiers()) {
+            if (auto subframeArchive = completeFrameArchive(subframeIdentifier))
+                archive->appendSubframeArchive(subframeArchive.releaseNonNull());
+        }
+
+        return archive;
+    }
+
+    ~LegacyWebArchiveCallbackAggregator()
+    {
+        if (m_callback)
```

Comment:
Can we ever create an aggregator with nullptr m_callback? I don't think we can???

Reply:
Yes, I will change it to assert.

Reply:
It seems default `CompletionHandler` does not have a function pointer

Reply:
What do you mean?

Reply:
nvm I was meant to say callers might pass in `CompletionHandler { }` in which case `!m_callback` will be true, but given that existing callers should always pass in a valid callback, I think assert is fine.